docs/adr/adr-012-daser-parallelization.md

```diff
@@ -0,0 +1,56 @@
+# ADR #012: DASer parallelization
+
+## Changelog
+
+- 2022-9-12: Started
+
+## Authors
+
+@walldiss
+
+## Context
+
+DAS is the process of verifying the availability of
+block data by sampling chunks or shares of those blocks. Package `das` implements an engine to confirm availability over the network's whole chain of blocks' data over the Availability interface.
+High availability of block data is a priority functionality for celestia-node. Its performance could benefit a lot from parallelization optimisation to make it able to fully utilise network bandwidth.
+
+## Previous approach
+
+Shares sampling, by its nature, is a network-bound operation that implies multiple network round-trips.
+The previous implementation of DAS'er used a single-thread synchronous approach.
+Meaning there was only one process sequentially performing sampling operations over past headers that were blocked by awaiting a response.
+
+## Decision
+
+Using multiple coordinated workers running in parallel drastically improves DAS'er performance by better network bandwidth utilization. On the downside, solution brings concurrency complexity.
+
+## Detailed Design
+
+To achieve parallelization DAS'er was split into core components:
+
+1. `Coordinator` holds current state of sampled headers and defines what headers should be sampled next.
+2. `Workers` perform sampling over a range of headers and communicate results back to the coordinator.
+3. `Subscriber` subscribes to updates of network head. If a new head is found, it will notify the `Coordinator`. Recent network head blocks will be prioritized for sampling to increase the availability of the most demanded blocks.
+4. `CheckpointStore` stores/loads`Coordinator` state as a checkpoint to allow seamless resume upon restart. `Coordinator` will store the state as a checkpoint on the exit and will resume sampling from the latest state.
+Also periodically stores checkpoint to storage in case node hard shutdown.
+
+![image](./img/daser-architecture-diagram.png)
+
+Sampling flow:
+
+1. `Coordinator` checks if worker concurrency limit is reached.
+2. If the limit is not reached, `Coordinator` forms a new sampling `Job`
+   1. Looks for priority `Job`.
+   2. If nothing is in priority, picks the next not sampled range for `Job`
```

Comment:
I do not fully understand this. So the next job is not the job with the next highest priority but priority is a binary property instead? Like either a job "is priority" or not?

Reply:
It was unclear so I've improved explanation by adding data type of priority - stack. Coordinator will pick jobs from priority until there is none.